browser/payments/payments_service_impl.cc

```diff
@@ -3,23 +3,267 @@
+  }
+
+  uint64_t request_id() override {
+    return request_id_;
   }
 
-  PaymentsServiceImpl::~PaymentsServiceImpl() {
+ private:
+  uint64_t request_id_;
+  net::URLFetcher* fetcher_;  // NOT OWNED
+};
+
+net::URLFetcher::RequestType URLMethodToRequestType(ledger::URL_METHOD method) {
+  switch(method) {
+    case ledger::URL_METHOD::GET:
+      return net::URLFetcher::RequestType::GET;
+    case ledger::URL_METHOD::POST:
+      return net::URLFetcher::RequestType::POST;
+    case ledger::URL_METHOD::PUT:
+      return net::URLFetcher::RequestType::PUT;
+    default:
+      NOTREACHED();
+      return net::URLFetcher::RequestType::GET;
   }
+}
+
+std::string LoadStateOnFileTaskRunner(
+    const base::FilePath& path) {
+  std::string data;
+  bool success = base::ReadFileToString(path, &data);
 
-  void PaymentsServiceImpl::CreateWallet() {
-    ledger_->createWallet();
+  // Make sure the file isn't empty.
+  if (!success || data.empty()) {
+    LOG(ERROR) << "Failed to read file: " << path.MaybeAsASCII();
+    return std::string();
   }
+  return data;
+}
+
+// `callback` has a WeakPtr so this won't crash if the file finishes
+// writing after PaymentServiceImpl has been destroyed
+void PostWriteCallback(
+    const base::Callback<void(bool success)>& callback,
+    scoped_refptr<base::SequencedTaskRunner> reply_task_runner,
+    bool write_success) {
+  // We can't run |callback| on the current thread. Bounce back to
+  // the |reply_task_runner| which is the correct sequenced thread.
+  reply_task_runner->PostTask(FROM_HERE,
+                              base::Bind(callback, write_success));
+}
+
+static uint64_t next_id = 1;
+
+}  // namespace
+
+PaymentsServiceImpl::PaymentsServiceImpl(Profile* profile) :
+    profile_(profile),
+    ledger_(ledger::Ledger::CreateInstance(this)),
+    file_task_runner_(base::CreateSequencedTaskRunnerWithTraits(
+        {base::MayBlock(), base::TaskPriority::BACKGROUND,
+         base::TaskShutdownBehavior::BLOCK_SHUTDOWN})),
+    ledger_state_path_(profile_->GetPath().Append("ledger_state")),
+    publisher_state_path_(profile_->GetPath().Append("publisher_state")) {
+}
+
+PaymentsServiceImpl::~PaymentsServiceImpl() {
+}
+
+void PaymentsServiceImpl::CreateWallet() {
+  ledger_->CreateWallet();
+}
+
+std::string PaymentsServiceImpl::GenerateGUID() const {
+  return base::GenerateGUID();
+}
+
+void PaymentsServiceImpl::Shutdown() {
+  fetchers_.clear();
+  ledger_.reset();
+  PaymentsService::Shutdown();
+}
+
+void PaymentsServiceImpl::OnWalletCreated(ledger::Result result) {
+  TriggerOnWalletCreated(result);
+}
+
+void PaymentsServiceImpl::OnReconcileComplete(ledger::Result result,
+                                              const std::string& viewing_id) {
+  LOG(ERROR) << "reconcile complete " << viewing_id;
+}
+
+void PaymentsServiceImpl::LoadLedgerState(
+    ledger::LedgerCallbackHandler* handler) {
+  base::PostTaskAndReplyWithResult(file_task_runner_.get(), FROM_HERE,
```

Comment:
static method and weakptr ensure that pending callbacks won't run on objects that have been destroyed